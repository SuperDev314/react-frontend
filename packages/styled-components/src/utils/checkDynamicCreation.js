// @flow

import { useRef } from 'react';

const invalidHookCallRe = /invalid hook call/i;
const seen = new Set();

export const checkDynamicCreation = (displayName: string, componentId?: string) => {
  if (process.env.NODE_ENV !== 'production') {
    const parsedIdString = componentId ? ` with the id of "${componentId}"` : '';
    const message =
      `The component ${displayName}${parsedIdString} has been created dynamically.\n` +
      "You may see this warning because you've called styled inside another component.\n" +
      'To resolve this only create new StyledComponents outside of any render method and function component.';

    try {
      // We purposefully call `useRef` outside of a component and expect it to throw
      // If it doesn't, then we're inside another component.
      // eslint-disable-next-line react-hooks/rules-of-hooks
      useRef();

      if (!seen.has(message)) {
        // eslint-disable-next-line no-console
        console.warn(message);
        seen.add(message);
      }
    } catch (error) {
      // The error here is expected, since we're expecting anything that uses `checkDynamicCreation` to
      // be called outside of a React component.
      if (invalidHookCallRe.test(error.message)) {
        // This shouldn't happen, but resets `warningSeen` if we had this error happen intermittently
        seen.delete(message);
      }
    }
  }
};
