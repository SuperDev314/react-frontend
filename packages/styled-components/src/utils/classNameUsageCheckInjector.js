// @flow
import ReactDOM from 'react-dom';
import getComponentName from './getComponentName';

const didWarnAboutClassNameUsage = new Set();

export default (target: Object) => {
  let elementClassName = '';

  const targetCDM = target.componentDidMount;

  // eslint-disable-next-line no-param-reassign
  target.componentDidMount = function componentDidMount() {
    if (typeof targetCDM === 'function') {
      targetCDM.call(this);
    }

    const forwardTarget = this.props.forwardedComponent.target;

    if (
      (target.props && target.props.suppressClassNameWarning) ||
      (target.attrs && target.attrs.suppressClassNameWarning) ||
      didWarnAboutClassNameUsage.has(forwardTarget)
    ) {
      return;
    }

    didWarnAboutClassNameUsage.add(forwardTarget);

    const classNames = elementClassName
      .replace(/ +/g, ' ')
      .trim()
      .split(' ');
    // eslint-disable-next-line react/no-find-dom-node
    const node: Element | null = (ReactDOM.findDOMNode(this): any);
    const selector = classNames.map(s => `.${s}`).join('');

    if (
      node &&
      node.nodeType === 1 &&
      !classNames.every(className => node.classList && node.classList.contains(className)) &&
      !node.querySelector(selector)
    ) {
      console.warn(
        `It looks like you've wrapped styled() around your React component (${getComponentName(
          forwardTarget
        )}), but the className prop is not being passed down to a child. No styles will be rendered unless className is composed within your React component.`
      );
    }
  };

  const prevRenderInner = target.renderInner;

  // eslint-disable-next-line no-param-reassign
  target.renderInner = function renderInner(...args) {
    const element = prevRenderInner.apply(this, args);

    elementClassName = element.props.className;

    return element;
  };
};
