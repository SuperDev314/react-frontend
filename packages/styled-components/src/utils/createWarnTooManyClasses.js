// @flow

export const LIMIT = 50;

export default (displayName: string) => {
  let generatedClasses = {};
  let warningSeen = false;

  return (className: string) => {
    if (!warningSeen) {
      generatedClasses[className] = true;
      if (Object.keys(generatedClasses).length >= LIMIT) {
        // Unable to find latestRule in test environment.
        /* eslint-disable no-console, prefer-template */
        console.warn(
          `Over ${LIMIT} classes were generated for component ${displayName}. This happens when some of the props you use for styling have many potential values and we need to make a new CSS class for each variant. Over time the stylesheet will grow and slow down your app.\n` +
            'For these particular CSS rules with high dynamicity, consider using the attrs() method together with a style object.\n' +
            'Example:\n' +
            '  const Component = styled.div.attrs(props => ({\n' +
            '    style: {\n' +
            '      background: props.background,\n' +
            '    },\n' +
            '  }))`width: 100%;`\n\n' +
            '  <Component />'
        );
        warningSeen = true;
        generatedClasses = {};
      }
    }
  };
};
