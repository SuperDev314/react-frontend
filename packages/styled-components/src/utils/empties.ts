export const EMPTY_ARRAY = (Object.freeze([]) as unknown) as Readonly<any[]>;
export const EMPTY_OBJECT = Object.freeze({});
