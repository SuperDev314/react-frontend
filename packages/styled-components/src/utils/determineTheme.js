// @flow
import { EMPTY_OBJECT } from './empties';

type Props = {
  theme?: any,
};

export default (props: Props, providedTheme: any, defaultProps: any = EMPTY_OBJECT) => {
  return (props.theme !== defaultProps.theme && props.theme) || providedTheme || defaultProps.theme;
};
