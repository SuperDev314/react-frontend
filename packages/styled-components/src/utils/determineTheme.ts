import { ExecutionProps } from '../types';
import { EMPTY_OBJECT } from './empties';

export default function determineTheme(
  props: ExecutionProps,
  providedTheme: any,
  defaultProps: any = EMPTY_OBJECT
) {
  return (props.theme && props.theme !== defaultProps.theme) || providedTheme || defaultProps.theme;
}
