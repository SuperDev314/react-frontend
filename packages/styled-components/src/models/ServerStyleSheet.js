// @flow
/* eslint-disable no-underscore-dangle */

import React from 'react';

import {
  IS_BROWSER,
  SC_ATTR,
  SC_VERSION_ATTR,
  SC_VERSION
} from '../constants';

import StyledError from '../utils/error';
import getNonce from '../utils/nonce';
import StyleSheet from '../sheet';
import StyleSheetManager from './StyleSheetManager';

declare var __SERVER__: boolean;

const CLOSING_TAG_R = /^\s*<\/[a-z]/i;

export default class ServerStyleSheet {
  sheet: StyleSheet;

  sealed: boolean;

  constructor() {
    this.sheet = new StyleSheet(true);
    this.sealed = false;
  }

  collectStyles(children: any) {
    if (this.sealed) {
      throw new StyledError(2);
    }

    this.sealed = true;
    return <StyleSheetManager sheet={this.sheet}>{children}</StyleSheetManager>;
  }

  getStyleTags = (): string => {
    const css = this.sheet.toString();
    const nonce = getNonce();
    const attrs = [
      nonce && `nonce="${nonce}"`,
      SC_ATTR,
      `${SC_VERSION_ATTR}="${SC_VERSION}"`
    ];

    const htmlAttr = attrs.filter(Boolean).join(' ');
    return `<style ${htmlAttr}>${css}</style>`;
  }

  getStyleElement = () => {
    const props = {
      [SC_ATTR]: '',
      [SC_VERSION_ATTR]: SC_VERSION,
      dangerouslySetInnerHTML: {
        __html: this.sheet.toString()
      }
    };

    const nonce = getNonce();
    if (nonce) {
      (props: any).nonce = nonce;
    }

    return <style {...props} />;
  }

  interleaveWithNodeStream(input: any) {
    if (!__SERVER__ || IS_BROWSER) {
      throw new StyledError(3);
    }

    const { Readable, Transform } = require('stream');
    const readableStream: Readable = input;
    const { sheet, getStyleTags } = this;

    const transformer = new Transform({
      transform: function appendStyleChunks(chunk, /* encoding */ _, callback) {
        // Get the chunk and retrieve the sheet's CSS as an HTML chunk,
        // then reset its rules so we get only new ones for the next chunk
        const renderedHtml = chunk.toString();
        const html = getStyleTags();
        sheet.clearTag();

        // prepend style html to chunk, unless the start of the chunk is a
        // closing tag in which case append right after that
        if (CLOSING_TAG_R.test(renderedHtml)) {
          const endOfClosingTag = renderedHtml.indexOf('>') + 1;
          const before = renderedHtml.slice(0, endOfClosingTag);
          const after = renderedHtml.slice(endOfClosingTag);

          this.push(before + html + after);
        } else {
          this.push(html + renderedHtml);
        }

        callback();
      },
    });

    readableStream.on('error', err => {
      // forward the error to the transform stream
      transformer.emit('error', err);
    });

    return readableStream.pipe(transformer);
  }
}
