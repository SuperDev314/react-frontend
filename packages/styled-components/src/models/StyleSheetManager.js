// @flow
import React, { createContext, Component, type Element } from 'react';
import PropTypes from 'prop-types';
import memoize from 'memoize-one';
import StyleSheet from './StyleSheet';
import ServerStyleSheet from './ServerStyleSheet';
import StyledError from '../utils/error';

type Props = {
  children?: Element<any>,
  sheet?: StyleSheet,
  target?: HTMLElement,
};

const StyleSheetContext = createContext();

export const StyleSheetConsumer = StyleSheetContext.Consumer;

export default class StyleSheetManager extends Component<Props> {
  static propTypes = {
    sheet: PropTypes.oneOfType([
      PropTypes.instanceOf(StyleSheet),
      PropTypes.instanceOf(ServerStyleSheet),
    ]),

    target: PropTypes.shape({
      appendChild: PropTypes.func.isRequired,
    }),
  };

  getContext: (sheet: ?StyleSheet, target: ?HTMLElement) => StyleSheet;

  constructor(props: Props) {
    super(props);
    this.getContext = memoize(this.getContext);
  }

  getContext(sheet: ?StyleSheet, target: ?HTMLElement) {
    if (sheet) {
      return sheet;
    } else if (target) {
      return new StyleSheet(target);
    } else {
      throw new StyledError(4);
    }
  }

  render() {
    const { children, sheet, target } = this.props;

    return (
      <StyleSheetContext.Provider value={this.getContext(sheet, target)}>
        {process.env.NODE_ENV !== 'production' ? React.Children.only(children) : children}
      </StyleSheetContext.Provider>
    );
  }
}
