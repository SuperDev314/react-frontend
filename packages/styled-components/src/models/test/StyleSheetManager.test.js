// @flow
/* eslint-disable react/prop-types */
import React from 'react';
import { renderToString } from 'react-dom/server';
import { render } from 'react-dom';
import TestRenderer from 'react-test-renderer';
import Frame, { FrameContextConsumer } from 'react-frame-component';
import stylisRTLPlugin from 'stylis-rtl';
import StyleSheetManager from '../StyleSheetManager';
import ServerStyleSheet from '../ServerStyleSheet';
import StyleSheet from '../../sheet';
import { resetStyled } from '../../test/utils';

let styled;
let consoleError;

const parallelWarning =
  'Warning: Detected multiple renderers concurrently rendering the same context provider. This is currently unsupported.';

describe('StyleSheetManager', () => {
  consoleError = console.error;

  beforeEach(() => {
    document.body.innerHTML = '';
    document.head.innerHTML = '';

    styled = resetStyled(true);

    jest
      .spyOn(console, 'error')
      .mockImplementation(msg => (msg !== parallelWarning ? consoleError(msg) : null));
  });

  it('should use given stylesheet instance', () => {
    const serverStyles = new ServerStyleSheet();
    const Title = styled.h1`
      color: palevioletred;
    `;
    renderToString(
      <StyleSheetManager sheet={serverStyles.sheet}>
        <Title />
      </StyleSheetManager>
    );
    expect(serverStyles.getStyleTags().includes(`palevioletred`)).toEqual(true);
  });

  it('should render its child', () => {
    const target = document.head;

    const Title = styled.h1`
      color: palevioletred;
    `;
    const renderedComp = TestRenderer.create(
      <StyleSheetManager target={target}>
        <Title />
      </StyleSheetManager>
    );

    expect(() => renderedComp.root.findByType(Title)).not.toThrowError();
  });

  it('should append style to given target', () => {
    const target = document.body;
    const Title = styled.h1`
      color: palevioletred;
    `;
    class Child extends React.Component {
      render() {
        return <Title />;
      }
    }

    expect(document.body.querySelectorAll('style')).toHaveLength(0);

    TestRenderer.create(
      <StyleSheetManager target={target}>
        <Child />
      </StyleSheetManager>
    );

    const styles = target.querySelector('style').textContent;

    expect(styles.includes(`palevioletred`)).toEqual(true);
  });

  it('should append style to given target in iframe', () => {
    const iframe = document.createElement('iframe');
    const app = document.createElement('div');

    document.body.appendChild(iframe);
    iframe.contentDocument.body.appendChild(app);

    const target = iframe.contentDocument.head;
    const Title = styled.h1`
      color: palevioletred;
    `;

    class Child extends React.Component {
      render() {
        return <Title />;
      }
    }

    render(
      <StyleSheetManager target={target}>
        <Child />
      </StyleSheetManager>,
      app
    );

    const styles = target.querySelector('style').textContent;
    expect(styles.includes(`palevioletred`)).toEqual(true);
  });

  it('should apply styles to appropriate targets for nested StyleSheetManagers', () => {
    const ONE = styled.h1`
      color: red;
    `;
    const TWO = styled.h2`
      color: blue;
    `;
    const THREE = styled.h3`
      color: green;
    `;

    TestRenderer.create(
      <div>
        <ONE />
        <StyleSheetManager target={document.head}>
          <div>
            <TWO />
            <StyleSheetManager target={document.body}>
              <THREE />
            </StyleSheetManager>
          </div>
        </StyleSheetManager>
      </div>
    );

    expect(document.head.innerHTML).toMatchSnapshot();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  // https://github.com/styled-components/styled-components/issues/1634
  it('should inject styles into two parallel contexts', async () => {
    const Title = styled.h1`
      color: palevioletred;
    `;

    // Injects the stylesheet into the document available via context
    const SheetInjector = ({ children, target }) => (
      <StyleSheetManager target={target}>{children}</StyleSheetManager>
    );

    class Child extends React.Component {
      componentDidMount() {
        const styles = this.props.document.querySelector('style').textContent;
        expect(styles.includes(`palevioletred`)).toEqual(true);
        this.props.resolve();
      }

      render() {
        return <Title />;
      }
    }

    const div = document.body.appendChild(document.createElement('div'));

    let promiseB;

    const promiseA = new Promise((resolveA, reject) => {
      promiseB = new Promise(resolveB => {
        try {
          // Render two iframes. each iframe should have the styles for the child injected into their head
          render(
            <div>
              <Frame>
                <FrameContextConsumer>
                  {({ document }) => (
                    <SheetInjector target={document.head}>
                      <Child document={document} resolve={resolveA} />
                    </SheetInjector>
                  )}
                </FrameContextConsumer>
              </Frame>
              <Frame>
                <FrameContextConsumer>
                  {({ document }) => (
                    <SheetInjector target={document.head}>
                      <Child document={document} resolve={resolveB} />
                    </SheetInjector>
                  )}
                </FrameContextConsumer>
              </Frame>
            </div>,
            div
          );
        } catch (e) {
          reject(e);
          div.parentElement.removeChild(div);
        }
      });
    });
    await Promise.all([promiseA, promiseB]);
    div.parentElement.removeChild(div);
  });

  it('should render styles in correct order when styled(StyledComponent) and StyleSheetManager is used', () => {
    const Red = styled.div`
      color: red;
    `;
    const RedChangedToBlue = styled(Red)`
      color: blue;
    `;
    const sheet = new StyleSheet();
    const App = () => (
      <StyleSheetManager sheet={sheet}>
        <RedChangedToBlue>I should be blue</RedChangedToBlue>
      </StyleSheetManager>
    );
    const attachPoint = document.body.appendChild(document.createElement('div'));
    render(<App />, attachPoint);
    // window.getComputedStyles would be perfect, but it seems that JSDOM
    // implementation of that function isn't complete, so need to work around
    // it.
    const source = document.documentElement.outerHTML;
    // regex in case test is run against minified CSS in the future
    const indexOfRedStyle = source.search('color:red');
    const indexOfBlueStyle = source.search('color:blue');
    expect(indexOfRedStyle).toBeGreaterThanOrEqual(0);
    expect(indexOfBlueStyle).toBeGreaterThanOrEqual(0);
    expect(indexOfBlueStyle).toBeGreaterThan(indexOfRedStyle);
  });

  it('passing stylis options via StyleSheetManager works', () => {
    const Test = styled.div`
      display: flex;
    `;

    TestRenderer.create(
      <StyleSheetManager stylisOptions={{ prefix: false }}>
        <Test>Foo</Test>
      </StyleSheetManager>
    );

    expect(document.head.innerHTML).toMatchInlineSnapshot(
      `"<style data-styled=\\"active\\" data-styled-version=\\"JEST_MOCK_VERSION\\">.sc-a{display:flex;}</style>"`
    );
  });

  it('passing stylis plugins via StyleSheetManager works', () => {
    const Test = styled.div`
      padding-left: 5px;
    `;

    TestRenderer.create(
      <StyleSheetManager stylusPlugins={[stylisRTLPlugin]}>
        <Test>Foo</Test>
      </StyleSheetManager>
    );

    expect(document.head.innerHTML).toMatchInlineSnapshot(
      `"<style data-styled=\\"active\\" data-styled-version=\\"JEST_MOCK_VERSION\\">.sc-a{padding-right:5px;}</style>"`
    );
  });

  it('nested StyleSheetManager with different injection modes works', () => {
    const Test = styled.div`
      padding-left: 5px;
    `;

    const Test2 = styled.div`
      background: red;
    `;

    const outerSheet = new StyleSheet({ useCSSOMInjection: true });

    TestRenderer.create(
      <StyleSheetManager sheet={outerSheet}>
        <div>
          <Test>Foo</Test>
          <StyleSheetManager disableCSSOMInjection>
            <Test2>Bar</Test2>
          </StyleSheetManager>
        </div>
      </StyleSheetManager>
    );

    expect(outerSheet.getTag().tag.getRule(0)).toMatchInlineSnapshot(
      `".sc-a {padding-left: 5px;}"`
    );

    expect(document.head.innerHTML).toMatchInlineSnapshot(
      `"<style data-styled=\\"active\\" data-styled-version=\\"JEST_MOCK_VERSION\\"></style><style data-styled=\\"active\\" data-styled-version=\\"JEST_MOCK_VERSION\\">.sc-b{background:red;}</style>"`
    );
  });
});
