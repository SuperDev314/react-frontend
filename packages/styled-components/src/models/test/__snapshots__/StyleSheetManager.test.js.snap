// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleSheetManager should apply styles to appropriate targets for nested StyleSheetManagers 1`] = `
"<style data-styled=\\"\\" data-styled-version=\\"JEST_MOCK_VERSION\\">
/* sc-component-id: sc-b */
.e{color:blue;}</style>"
`;

exports[`StyleSheetManager should apply styles to appropriate targets for nested StyleSheetManagers 2`] = `
"<style data-styled=\\"\\" data-styled-version=\\"JEST_MOCK_VERSION\\">
/* sc-component-id: sc-c */
.f{color:green;}</style>"
`;

exports[`StyleSheetManager ssr should extract CSS outside the nested StyleSheetManager 1`] = `"<div><h1 class=\\"sc-a c\\"></h1><div></div></div>"`;

exports[`StyleSheetManager ssr should extract CSS outside the nested StyleSheetManager 2`] = `
"<style data-styled=\\"c\\" data-styled-version=\\"JEST_MOCK_VERSION\\">
/* sc-component-id: sc-a */
.c{color:red;}</style>"
`;
